File: app/src/main/java/de/squiray/dailylist/DailylistApp.java
package de.squiray.dailylist;

import android.app.Activity;
import android.app.Application;
import android.os.Build;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasActivityInjector;
import de.squiray.dailylist.presentation.di.component.DaggerAppComponent;
import de.squiray.dailylist.presentation.service.DailyStrikeResetAlarm;
import de.squiray.dailylist.presentation.service.DailyTodoLimitResetAlarm;
import de.squiray.dailylist.util.logging.CrashLogging;
import de.squiray.dailylist.util.logging.DebugLogger;
import de.squiray.dailylist.util.logging.ReleaseLogger;
import timber.log.Timber;

public class DailylistApp extends Application implements HasActivityInjector {

    @Inject
    DispatchingAndroidInjector<Activity> dispatchingActivityInjector;

    @Inject
    DailyTodoLimitResetAlarm dailyTodoLimitResetAlarm;

    @Inject
    DailyStrikeResetAlarm dailyStrikeResetAlarm;

    @Override
    public void onCreate() {
        super.onCreate();
        initDagger();
        setupLogging();
        logAppStats();
        dailyTodoLimitResetAlarm.scheduleAlarm();
        dailyStrikeResetAlarm.scheduleAlarm();
    }

    private void initDagger() {
        DaggerAppComponent.builder()
                .application(this)
                .build()
                .inject(this);
    }

    private void setupLogging() {
        setupLoggingFramework();
        CrashLogging.setup();
    }

    private void setupLoggingFramework() {
        if (BuildConfig.DEBUG) {
            Timber.plant(new DebugLogger());
        }
        Timber.plant(new ReleaseLogger(this));
    }

    private void logAppStats() {
        Timber.tag("App").i("DailyTodo v%s (%d) started on android %s / API%d using a %s", //
                BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE, //
                Build.VERSION.RELEASE, Build.VERSION.SDK_INT, //
                Build.MODEL);
    }

    @Override
    public AndroidInjector<Activity> activityInjector() {
        return dispatchingActivityInjector;
    }
}

File: app/src/main/java/de/squiray/dailylist/DailyListApp.kt
package de.squiray.dailylist

import android.content.Intent
import android.os.Build
import dagger.android.AndroidInjector
import dagger.android.support.DaggerApplication
import de.squiray.dailylist.presentation.di.component.DaggerAppComponent
import de.squiray.dailylist.presentation.service.DailyListService
import de.squiray.dailylist.util.logging.CrashLogging
import de.squiray.dailylist.util.logging.DebugLogger
import de.squiray.dailylist.util.logging.ReleaseLogger
import timber.log.Timber

class DailyListApp : DaggerApplication() {

    override fun onCreate() {
        super.onCreate()
        setupLogging()
        logAppStats()
        startDailyListService()
    }

    override fun applicationInjector(): AndroidInjector<out DaggerApplication> {
        return DaggerAppComponent
                .builder()
                .application(this)
                .build()
    }

    private fun startDailyListService() {
        startService(Intent(this, DailyListService::class.java))
    }

    private fun setupLogging() {
        setupLoggingFramework()
        CrashLogging.setup()
    }

    private fun setupLoggingFramework() {
        if (BuildConfig.DEBUG) {
            Timber.plant(DebugLogger())
        }
        Timber.plant(ReleaseLogger(this))
    }

    private fun logAppStats() {
        Timber.tag("App").i("DailyTodo v%s (%d) started on android %s / API%d using a %s", //
                BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE, //
                Build.VERSION.RELEASE, Build.VERSION.SDK_INT, //
                Build.MODEL)
    }
}
